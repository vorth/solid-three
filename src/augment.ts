import { S3 } from "./";

export const $S3C = Symbol("solid-three");

/**
 * A utility to augment a `three` instance with additional data.
 * This data can be accessed behind the `S3C` symbol and is used internally in `solid-three`.
 *
 * @param instance - `three` instance
 * @param augmentation - additional data: `{ props }`
 * @returns the `three` instance with the additional data
 */
export const augment = <T>(instance: T, augmentation: any) => {
  instance[$S3C] = { children: new Set(), ...augmentation };
  return instance as T & {
    [$S3C]: S3.Metadata<T>;
  };
};
