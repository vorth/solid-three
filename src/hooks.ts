import { Accessor, Resource, createContext, createResource, useContext } from "solid-js";
import { ThreeContext } from "./types";

/**
 * Custom hook to access all necessary Three.js objects needed to manage a 3D scene.
 * This hook must be used within a component that is a descendant of the `<Canvas/>` component.
 *
 * @template T The expected return type after applying the callback to the context.
 * @param {Function} [callback] - Optional callback function that processes and returns a part of the context.
 * @returns {ThreeContext | Accessor<T>} Returns the ThreeContext directly, or an accessor if a callback is provided.
 * @throws {Error} Throws an error if used outside of the Canvas component context.
 */
export function useThree(): ThreeContext;
export function useThree<T>(callback: (value: ThreeContext) => T): Accessor<T>;
export function useThree(callback?: (value: ThreeContext) => any) {
  const store = useContext(threeContext);
  if (!store) {
    throw new Error("S3F: Hooks can only be used within the Canvas component!");
  }
  if (callback) return () => callback(store);
  return store;
}
export const threeContext = createContext<ThreeContext>(null!);

/**
 * Hook to register a callback that will be executed on each animation frame within the `<Canvas/>` component.
 * This hook must be used within a component that is a descendant of the `<Canvas/>` component.
 *
 * @param callback - The callback function to be executed on each frame.
 * @throws {Error} Throws an error if used outside of the Canvas component context.
 */
export const useFrame = (
  callback: (context: ThreeContext, delta: number, frame?: XRFrame) => void,
) => {
  const addFrameListener = useContext(frameContext);
  if (!addFrameListener) {
    throw new Error("S3F: Hooks can only be used within the Canvas component!");
  }
  addFrameListener(callback);
};
export const frameContext =
  createContext<
    (callback: (context: ThreeContext, delta: number, frame?: XRFrame) => void) => void
  >();

/**
 * Hook to create and manage a resource using a Three.js loader. It ensures that the loader is
 * reused if it has been instantiated before, and manages the resource lifecycle automatically.
 *
 * @template TResult The type of the resolved data when the loader completes loading.
 * @template TArg The argument type expected by the loader function.
 * @param Constructor - The loader class constructor.
 * @param args - The arguments to be passed to the loader function, wrapped in an accessor to enable reactivity.
 * @returns An accessor containing the loaded resource, re-evaluating when inputs change.
 */
export const useLoader = <TArg extends string | readonly string[], TLoader extends Loader>(
  Constructor: new () => TLoader,
  args: Accessor<TArg>,
  setup?: (loader: TLoader) => void,
) => {
  let cache = LOADER_CACHE.get(Constructor) as LoaderCache<TLoader>;
  if (!cache) {
    cache = {
      loader: new Constructor(),
      resources: {},
    };
    LOADER_CACHE.set(Constructor, cache);
  }
  const { loader, resources } = cache;
  setup?.(loader);

  const load = (arg: string) => {
    if (resources[arg]) return resources[arg];
    return (resources[arg] = new Promise((resolve, reject) =>
      loader.load(
        arg,
        value => {
          resources[arg] = value;
          resolve(value);
        },
        undefined,
        reject,
      ),
    ));
  };

  const [resource] = createResource(args, args =>
    Array.isArray(args)
      ? Promise.all((args as string[]).map(arg => load(arg)))
      : load(args as string),
  );

  return resource as Resource<
    UseLoaderOverload<string, TLoader extends Loader<infer U> ? U : never, TArg>
  >;
};
type Loader<TLoaderResult = any> = {
  load: (
    value: string,
    onLoad: (value: TLoaderResult) => void,
    onProgress: (() => void) | undefined,
    onReject: ((error: ErrorEvent) => void) | undefined,
  ) => void | null;
};
type UseLoaderOverload<TLoaderArg, TLoaderResult, TArg> = TArg extends readonly TLoaderArg[]
  ? { [K in keyof TArg]: TLoaderResult }
  : TLoaderResult;

const LOADER_CACHE = new Map<any, LoaderCache>();
type LoaderCache<T = Loader<any>> = { loader: T; resources: {} };
