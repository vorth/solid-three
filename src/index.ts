export { $S3C } from "./augment";
export { Canvas } from "./canvas";
export { useFrame, useLoader, useThree } from "./hooks";
export { T, extend } from "./proxy";
export * from "./types";
export { buildGraph } from "./utils/build-graph";
