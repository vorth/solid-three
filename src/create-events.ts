import { onCleanup } from "solid-js";
import { Intersection, Object3D } from "three";
import { S3 } from "./";
import { $S3C } from "./augment";
import { isAugmentedElement } from "./utils/is-augmented-element";
import { removeElementFromArray } from "./utils/remove-element-from-array";

/**
 * Checks if a given string is a valid event type within the system.
 *
 * @param type - The type of the event to check.
 * @returns `true` if the type is a recognized `EventType`, otherwise `false`.
 */
export const isEventType = (type: string): type is S3.EventName =>
  /^on(Pointer|Click|DoubleClick|ContextMenu|Wheel|Mouse)/.test(type);

/**
 * Initializes and manages event handling for `AugmentedElement<Object3D>`.
 *
 * @param context
 * @returns An object containing `addEventListener`-function and `eventRegistry`-object.
 */
export const createEvents = (context: S3.Context) => {
  /**
   * An object keeping track of all the `AugmentedElement<Object3D>` that are listening to a specific event.
   */
  const eventRegistry = {
    onMouseMove: [] as S3.Instance<Object3D>[],
    onMouseUp: [] as S3.Instance<Object3D>[],
    onMouseDown: [] as S3.Instance<Object3D>[],
    onPointerMove: [] as S3.Instance<Object3D>[],
    onPointerUp: [] as S3.Instance<Object3D>[],
    onPointerDown: [] as S3.Instance<Object3D>[],
    onWheel: [] as S3.Instance<Object3D>[],
    onClick: [] as S3.Instance<Object3D>[],
    onDoubleClick: [] as S3.Instance<Object3D>[],
  } as const;

  // Creates a `ThreeEvent` from the current `MouseEvent` | `WheelEvent`.
  const createThreeEvent = <TEvent extends MouseEvent | WheelEvent>(
    nativeEvent: TEvent,
  ): S3.Event<TEvent> => {
    const event = {
      ...nativeEvent,
      nativeEvent,
      stopped: false,
      stopPropagation: () => (event.stopped = true),
    };
    return event;
  };

  // Performs a raycast from the camera through the mouse position to find intersecting 3D objects.
  const raycast = <TNativeEvent extends MouseEvent | WheelEvent>(
    nativeEvent: TNativeEvent,
    type: keyof typeof eventRegistry,
  ): Intersection<S3.Instance<Object3D>>[] => {
    context.setPointer(pointer => {
      pointer.x = (nativeEvent.offsetX / window.innerWidth) * 2 - 1;
      pointer.y = -(nativeEvent.offsetY / window.innerHeight) * 2 + 1;
      return pointer;
    });

    context.raycaster.setFromCamera(context.pointer, context.camera);

    const duplicates = new Set<S3.Instance<Object3D>>();

    // NOTE:  we currently perform a recursive intersection-test just as r3f.
    //        this method performs a lot of duplicate intersection-tests.
    const intersections: Intersection<S3.Instance<Object3D>>[] = context.raycaster.intersectObjects(
      eventRegistry[type],
      true,
    );

    return (
      intersections
        // sort by distance
        .sort((a, b) => a.distance - b.distance)
        // remove duplicates
        .filter(({ object }) => {
          if (duplicates.has(object)) return false;
          duplicates.add(object);
          return true;
        })
    );
  };

  // Propagates an event down through the ancestors of a given `Object3D` in a scene graph,
  // calling the event handler for each ancestor as long as the event has not been marked as stopped.
  const bubbleDown = <
    TNativeEvent extends MouseEvent | WheelEvent,
    TEvent extends S3.Event<TNativeEvent>,
  >(
    element: S3.Instance<Object3D>,
    type: S3.EventName,
    event: TEvent,
  ) => {
    let node: Object3D | null = element.parent;
    while (node) {
      // If event has been stopped with event.stopPropagation() we break out.
      if (event.stopped) break;
      // If node is an AugmentedElement we call the type's event-handler if it is defined.
      if (isAugmentedElement(node)) {
        node[$S3C].props[type]?.(event);
      }
      // We bubble a layer down.
      node = node.parent;
    }
  };

  // A handler-factory for `on{Pointer|Mouse}Move` events.
  // This handler manages also its derived events:
  // - `on{Pointer|Mouse}Enter`
  // - `on{Pointer|Mouse}Leave`
  const createMoveHandler = (type: "Pointer" | "Mouse") => (nativeEvent: MouseEvent) => {
    const moveEvent = createThreeEvent(nativeEvent);
    const enterEvent = createThreeEvent(nativeEvent);

    let staleIntersects = new Set(priorIntersects[type]);

    for (const intersection of raycast(nativeEvent, `on${type}Move`)) {
      const props = intersection.object[$S3C].props;

      if (!enterEvent.stopped && !priorIntersects[type].has(intersection.object)) {
        props[`on${type}Enter`]?.(enterEvent);
        bubbleDown(intersection.object, `on${type}Enter`, enterEvent);
      }

      if (!moveEvent.stopped) {
        props[`on${type}Move`]?.(moveEvent);
        bubbleDown(intersection.object, `on${type}Move`, moveEvent);
      }

      staleIntersects.delete(intersection.object);
      priorIntersects[type].add(intersection.object);

      if (moveEvent.stopped && enterEvent.stopped) break;
    }

    if (priorMoveEvents[type]) {
      const leaveEvent = createThreeEvent(priorMoveEvents[type]!);

      for (const object of staleIntersects.values()) {
        priorIntersects[type].delete(object);

        if (!leaveEvent.stopped) {
          const props = object[$S3C].props;
          props[`on${type}Leave`]?.(leaveEvent);
          bubbleDown(object, `on${type}Leave`, leaveEvent);
        }
      }
    }

    priorMoveEvents[type] = nativeEvent;
  };
  const priorIntersects = {
    Mouse: new Set<S3.Instance<Object3D>>(),
    Pointer: new Set<S3.Instance<Object3D>>(),
  };
  const priorMoveEvents = {
    Mouse: undefined as undefined | MouseEvent,
    Pointer: undefined as undefined | MouseEvent,
  };

  // Creates a generic event handler for events other than `move` and its derived events.
  const createEventHandler =
    <TEvent extends MouseEvent | WheelEvent>(type: keyof typeof eventRegistry) =>
    (nativeEvent: TEvent) => {
      const event = createThreeEvent(nativeEvent);
      for (const { object } of raycast(nativeEvent, type)) {
        object[$S3C].props[type]?.(event);
        bubbleDown(object, type, event);
        if (event.stopped) break;
      }
    };

  // Register event handlers to the canvas
  context.canvas.addEventListener("mousemove", createMoveHandler("Mouse"));
  context.canvas.addEventListener("pointermove", createMoveHandler("Pointer"));

  context.canvas.addEventListener("mousedown", createEventHandler("onMouseDown"));
  context.canvas.addEventListener("pointerdown", createEventHandler("onPointerDown"));

  context.canvas.addEventListener("mouseup", createEventHandler("onMouseUp"));
  context.canvas.addEventListener("pointerup", createEventHandler("onPointerUp"));

  context.canvas.addEventListener("wheel", createEventHandler("onWheel"));
  context.canvas.addEventListener("click", createEventHandler("onClick"));
  context.canvas.addEventListener("dblclick", createEventHandler("onDoubleClick"));

  /**
   * Registers an `AugmentedElement<Object3D>` with the event handling system.
   *
   * @param object - The 3D object to register.
   * @param type - The type of event the object should listen for.
   */
  const addEventListener = (object: S3.Instance<Object3D>, type: S3.EventName) => {
    const isDerivedEvent = type.includes("Enter") || type.includes("Leave");
    const isPointerEvent = type.includes("Pointer");

    // Derived events are handled by `on{Pointer|Mouse}Move`
    const derivedType = isDerivedEvent ? `on${isPointerEvent ? "Pointer" : "Mouse"}Move` : type;

    if (!eventRegistry[derivedType].find(value => value === object)) {
      eventRegistry[derivedType].push(object);
    }

    onCleanup(() => {
      // NOTE:  When a move/derived event-handler cleans up, it only removes the object from the registry
      //        if the object is currently not listening to another move/derived-event.
      if (derivedType.includes("Move")) {
        const props = object[$S3C].props;
        if (isPointerEvent) {
          if ("onPointerMove" in props || "onPointerEnter" in props || "onPointerLeave" in props) {
            return;
          }
        } else {
          if ("onMouseMove" in props || "onMouseEnter" in props || "onMouseLeave" in props) {
            return;
          }
        }
      }

      removeElementFromArray(eventRegistry[type], object);
    });
  };

  return { addEventListener, eventRegistry };
};
