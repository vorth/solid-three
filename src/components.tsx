import { JSXElement, ParentProps, createMemo, createRenderEffect, mergeProps } from "solid-js";
import { JSX } from "solid-js/jsx-runtime";
import { Object3D } from "three";
import { augment } from "./augment";
import { threeContext, useThree } from "./hooks";
import { manageProps, manageSceneGraph } from "./props";
import { AugmentedElement, ThreeElement, ThreeProps } from "./types";
import { isAugmentedElement } from "./utils/is-augmented-element";
import { withContext } from "./utils/with-context";

/**
 * A component for placing its children outside the regular `solid-three` scene graph managed by Solid's reactive system.
 * This is useful for bypassing the normal rendering flow and manually managing children, similar to Solid's Portal but specific to `solid-three`.
 *
 * @function Portal
 * @param {PortalProps} props - The component props containing `children` to be rendered and an optional Object3D `element` to be rendered into.
 * @returns {JSX.Fragment} An empty JSX element.
 */
export const Portal = (props: PortalProps) => {
  const context = useThree();
  const scene = createMemo(() =>
    props.element
      ? isAugmentedElement(props.element)
        ? props.element
        : augment(props.element, { props: {} })
      : context.scene,
  );

  createRenderEffect(() => {
    manageSceneGraph(
      scene(),
      withContext(
        () => props.children as unknown as AugmentedElement | AugmentedElement[],
        threeContext,
        mergeProps(context, {
          get scene() {
            return scene();
          },
        }),
      ),
    );
  });
  return <></>;
};
type PortalProps = ParentProps<{ element?: ThreeElement<Object3D> | AugmentedElement<Object3D> }>;

/**
 * Wraps a `Three.js` object instance and allows it to be used as a JSX component within a `solid-three` scene.
 *
 * @function Primitive
 * @template T - Extends ThreeElement which includes types from Three.js (like Mesh, Light, etc.).
 * @param {PrimitiveProps<T>} props - The properties for the Three.js object including the object instance's methods,
 *                                    optional children, and a ref that provides access to the object instance.
 * @returns {JSX.Element} The Three.js object wrapped as a JSX element, allowing it to be used within Solid's component system.
 */
export function Primitive<T extends ThreeElement>(props: PrimitiveProps<T>) {
  const memo = createMemo(() => augment(props.object, { props }));

  manageProps(memo, props);

  return memo as unknown as JSX.Element;
}
type PrimitiveProps<T> = Omit<ThreeProps<T>, "object" | "children" | "ref" | "args"> & {
  object: T;
  children?: JSXElement;
  ref?: T | ((value: T) => void);
};
