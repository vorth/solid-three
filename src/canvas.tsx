import { createResizeObserver } from "@solid-primitives/resize-observer";
import { ComponentProps, JSX, createRenderEffect, splitProps } from "solid-js";
import { OrthographicCamera, PerspectiveCamera, Raycaster, Scene, WebGLRenderer } from "three";
import { createThree } from "./create-three";
import { ThreeProps } from "./types";

export interface CanvasProps extends ComponentProps<"div"> {
  camera?: Partial<ThreeProps<PerspectiveCamera> | ThreeProps<OrthographicCamera>>;
  fallback?: JSX.Element;
  gl?: Partial<ThreeProps<WebGLRenderer>> | ((canvas: HTMLCanvasElement) => WebGLRenderer);
  orthographic?: boolean;
  raycaster?: Partial<ThreeProps<Raycaster>>;
  scene?: Partial<ThreeProps<Scene>>;
  style?: JSX.CSSProperties;
  shadows?: boolean | "basic" | "percentage" | "soft" | "variance" | WebGLRenderer["shadowMap"];
  linear?: boolean;
  flat?: boolean;
  frameloop?: "never" | "demand" | "always";
}

/**
 * Serves as the root component for all 3D scenes created with `solid-three`. It initializes
 * the Three.js rendering context, including a WebGL renderer, a scene, and a camera.
 * All `<T/>`-components must be children of this Canvas. Hooks such as `useThree` and
 * `useFrame` should only be used within this component to ensure proper context.
 *
 * @function Canvas
 * @param {CanvasProps} props - Configuration options include camera settings, style, and children elements.
 * @returns {JSX.Element} A div element containing the WebGL canvas configured to occupy the full available space.
 */
export function Canvas(_props: CanvasProps) {
  const [props, canvasProps] = splitProps(_props, ["fallback", "camera", "children", "ref"]);

  const canvas = (<canvas style={{ width: "100%", height: "100%" }} />) as HTMLCanvasElement;
  const container = (
    <div style={{ width: "100%", height: "100%" }}>{canvas}</div>
  ) as HTMLDivElement;

  const context = createThree(canvas, props);

  // Resize observer for the canvas to adjust camera and renderer on size change
  createResizeObserver(
    () => container,
    () => {
      context.gl.setSize(window.innerWidth, window.innerHeight);
      context.gl.setPixelRatio(window.devicePixelRatio);

      if (context.camera instanceof OrthographicCamera) {
        context.camera.left = window.innerWidth / -2;
        context.camera.right = window.innerWidth / 2;
        context.camera.top = window.innerHeight / 2;
        context.camera.bottom = window.innerHeight / -2;
      } else {
        context.camera.aspect = window.innerWidth / window.innerHeight;
      }
      context.camera.updateProjectionMatrix();
    },
  );

  // Assign ref
  createRenderEffect(() => {
    if (props.ref instanceof Function) props.ref(container);
    else props.ref = container;
  });

  return (
    <div
      {...canvasProps}
      style={{
        position: "relative",
        width: "100%",
        height: "100%",
        overflow: "hidden",
        ...canvasProps.style,
      }}
    >
      {container}
    </div>
  );
}
