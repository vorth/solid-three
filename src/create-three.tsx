import {
  Accessor,
  children,
  createMemo,
  createRenderEffect,
  createSignal,
  mergeProps,
  onCleanup,
} from "solid-js";
import {
  ACESFilmicToneMapping,
  BasicShadowMap,
  Camera,
  LinearEncoding,
  NoToneMapping,
  OrthographicCamera,
  PCFShadowMap,
  PCFSoftShadowMap,
  PerspectiveCamera,
  Raycaster,
  Scene,
  VSMShadowMap,
  Vector2,
  WebGLRenderer,
  sRGBEncoding,
} from "three";
import { augment } from "./augment";
import { CanvasProps } from "./canvas";
import { createEvents } from "./create-events";
import { frameContext, threeContext } from "./hooks";
import { canvasPropsContext, eventContext } from "./internal-context";
import { manageProps, manageSceneGraph } from "./props";
import { AugmentedElement, ThreeContext } from "./types";
import { defaultProps } from "./utils/default-props";
import { removeElementFromArray } from "./utils/remove-element-from-array";
import { withMultiContexts } from "./utils/with-context";

/**
 * Creates and manages a `solid-three` scene. It initializes necessary objects like
 * camera, renderer, raycaster, and scene, manages the scene graph, setups up an event system
 * and rendering loop based on the provided properties.
 *
 * @param {HTMLCanvasElement} canvas - The HTML canvas element on which Three.js will render.
 * @param {CanvasProps} props - Configuration properties.
 * @returns - a `ThreeContext` with additional properties including eventRegistry and addFrameListener.
 */
export function createThree(canvas: HTMLCanvasElement, props: CanvasProps) {
  const canvasProps = defaultProps(props, { frameloop: "always" });

  const [pointer, setPointer] = createSignal(new Vector2(), { equals: false });
  const frameListeners: ((context: ThreeContext, delta: number, frame?: XRFrame) => void)[] = [];

  // Adds a callback to be called on each frame
  const addFrameListener = (
    callback: (context: ThreeContext, delta: number, frame?: XRFrame) => void,
  ) => {
    frameListeners.push(callback);
    const cleanup = () => removeElementFromArray(frameListeners, callback);
    onCleanup(cleanup);
    return cleanup;
  };

  // Create core elements
  const { camera, gl, raycaster, scene } = createCoreElements(canvas, canvasProps);

  // Handle frame behavior in WebXR
  const handleXRFrame: XRFrameRequestCallback = (timestamp: number, frame?: XRFrame) => {
    if (canvasProps.frameloop === "never") return;
    render(timestamp, frame);
  };
  // Toggle render switching on session
  const handleSessionChange = () => {
    context.gl.xr.enabled = context.gl.xr.isPresenting;
    context.gl.xr.setAnimationLoop(context.gl.xr.isPresenting ? handleXRFrame : null);
  };
  // WebXR session-manager
  const xr = {
    connect() {
      context.gl.xr.addEventListener("sessionstart", handleSessionChange);
      context.gl.xr.addEventListener("sessionend", handleSessionChange);
    },
    disconnect() {
      context.gl.xr.removeEventListener("sessionstart", handleSessionChange);
      context.gl.xr.removeEventListener("sessionend", handleSessionChange);
    },
  };

  // Render-functions
  let isRenderPending = false;
  const render = (timestamp: number, frame?: XRFrame) => {
    isRenderPending = false;
    context.gl.render(context.scene, context.camera);
    frameListeners.forEach(listener => listener(context, timestamp, frame));
  };
  const requestRender = () => {
    if (isRenderPending) return;
    isRenderPending = true;
    requestAnimationFrame(render);
  };

  // Compose three-context
  const context: ThreeContext = {
    canvas,
    // Add core elements
    get camera() {
      return camera();
    },
    get gl() {
      return gl();
    },
    get raycaster() {
      return raycaster();
    },
    get scene() {
      return scene();
    },
    // Current normalized, centric pointer coordinates
    get pointer() {
      return pointer();
    },
    setPointer,
    render,
    requestRender,
    xr,
  };

  // Manage core elements
  withMultiContexts(
    () => manageCoreElements(canvasProps, context),
    [
      [threeContext, context],
      [canvasPropsContext, canvasProps],
    ],
  );

  // Initialize event-system
  const { addEventListener, eventRegistry } = createEvents(context);

  // Manage scene-graph
  manageSceneGraph(
    context.scene,
    children(
      withMultiContexts(
        () => canvasProps.children,
        [
          // Dependency Injection
          [threeContext, context],
          [frameContext, addFrameListener],
          [eventContext, addEventListener],
          [canvasPropsContext, canvasProps],
        ],
      ),
    ) as unknown as Accessor<AugmentedElement[]>,
  );

  // Render-loop
  const loop = (value: number) => {
    if (canvasProps.frameloop === "always") {
      requestAnimationFrame(loop);
      context.render(value);
    }
  };
  createRenderEffect(() => {
    if (canvasProps.frameloop === "always") {
      requestAnimationFrame(loop);
    }
  });

  return mergeProps(context, { eventRegistry, addFrameListener });
}

/**
 * Initializes the core components of a `solid-three` rendering context. This function creates
 * essential elements like the camera, scene, raycaster, and WebGL renderer that are required
 * to setup a fully functional `solid-three` environment.
 *
 * @param {HTMLCanvasElement} canvas - The HTML canvas element to be used for the WebGL renderer.
 * @param {CanvasProps} props - Configuration properties that define specifics such as camera type,
 *                              scene configuration, raycaster parameters, and renderer options.
 * @returns - Returns objects providing reactive access to the camera, WebGL renderer, raycaster,
 *            and scene, allowing these elements to be integrated into the Solid.js reactive system.
 */
const createCoreElements = (canvas: HTMLCanvasElement, props: CanvasProps) => ({
  camera: createMemo(() =>
    augment(
      props.camera instanceof Camera
        ? (props.camera as OrthographicCamera | PerspectiveCamera)
        : props.orthographic
        ? new OrthographicCamera()
        : new PerspectiveCamera(),
      {
        get props() {
          return props.camera || {};
        },
      },
    ),
  ),
  scene: createMemo(() =>
    augment(props.scene instanceof Scene ? props.scene : new Scene(), {
      get props() {
        return props.scene || {};
      },
    }),
  ),
  raycaster: createMemo(() =>
    augment(props.raycaster instanceof Raycaster ? props.raycaster : new Raycaster(), {
      get props() {
        return props.raycaster || {};
      },
    }),
  ),
  gl: createMemo(() =>
    augment(
      props.gl instanceof WebGLRenderer
        ? // props.gl can be a WebGLRenderer provided by the user
          props.gl
        : typeof props.gl === "function"
        ? // or a callback that returns a Renderer
          props.gl(canvas)
        : // if props.gl is not defined we default to a WebGLRenderer
          new WebGLRenderer({ canvas }),
      {
        get props() {
          return props.gl || {};
        },
      },
    ),
  ),
});

/**
 * Manages and updates the properties of core Three.js components such as the camera, scene, and
 * WebGL renderer based on external properties. This function applies reactive effects to update
 * these components dynamically as their configurations change. It handles details like shadow mapping,
 * color management, and tone mapping to align with the provided properties.
 *
 * @param {CanvasProps} props - The properties containing configuration updates for the camera,
 *                              scene, renderer, and other related settings.
 * @param {ThreeContext} context - The current Three.js context that contains the core elements
 *                                 (camera, scene, raycaster, renderer) which will be managed.
 */
export const manageCoreElements = (props: CanvasProps, context: ThreeContext) => {
  // Manage context.camera
  createRenderEffect(() => {
    if (!props.camera || props.camera instanceof Camera) return;
    manageProps(() => context.camera, props.camera);
    // NOTE:  Manually update camera's matrix with updateMatrixWorld is needed.
    //        Otherwise casting a ray immediately after start-up will cause the incorrect matrix to be used.
    context.camera.updateMatrixWorld(true);
  });
  // Manage context.scene
  createRenderEffect(() => {
    if (!props.scene || props.scene instanceof Scene) return;
    manageProps(() => context.scene, props.scene);
  });
  // Manage context.gl
  createRenderEffect(() => {
    // Set shadow-map
    createRenderEffect(() => {
      if (context.gl.shadowMap) {
        const oldEnabled = context.gl.shadowMap.enabled;
        const oldType = context.gl.shadowMap.type;
        context.gl.shadowMap.enabled = !!props.shadows;

        if (typeof props.shadows === "boolean") {
          context.gl.shadowMap.type = PCFSoftShadowMap;
        } else if (typeof props.shadows === "string") {
          const types = {
            basic: BasicShadowMap,
            percentage: PCFShadowMap,
            soft: PCFSoftShadowMap,
            variance: VSMShadowMap,
          };
          context.gl.shadowMap.type = types[props.shadows] ?? PCFSoftShadowMap;
        } else if (typeof props.shadows === "object") {
          Object.assign(context.gl.shadowMap, props.shadows);
        }

        if (oldEnabled !== context.gl.shadowMap.enabled || oldType !== context.gl.shadowMap.type)
          context.gl.shadowMap.needsUpdate = true;
      }
    });
    // Color management and tone-mapping
    manageProps(() => context.gl, {
      get outputEncoding() {
        return props.linear ? LinearEncoding : sRGBEncoding;
      },
      get toneMapping() {
        return props.flat ? NoToneMapping : ACESFilmicToneMapping;
      },
    });
    // Connect to xr if property exists
    if (context.gl.xr) context.xr.connect();
    // Manage props
    if (!props.gl || props.gl instanceof WebGLRenderer) return;
    manageProps(() => context.gl, props.gl);
  });
  // Manage context.raycaster
  createRenderEffect(() => {
    if (!props.raycaster || props.raycaster instanceof Raycaster) return;
    manageProps(() => context.raycaster, props.raycaster);
  });
};
