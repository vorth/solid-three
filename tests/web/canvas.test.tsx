// use default export for jest.spyOn
import { render } from "@solidjs/testing-library";

import { describe, expect, it } from "vitest";
import { T } from "../../src";
import { TestCanvas } from "../../src/testing";

describe("web Canvas", () => {
  it("should correctly mount", async () => {
    let renderer = render(() => (
      <TestCanvas>
        <T.Group />
      </TestCanvas>
    ));

    expect(renderer.container).toMatchSnapshot();
  });

  it("should forward ref", async () => {
    let ref: HTMLDivElement;

    render(() => (
      <TestCanvas ref={ref}>
        <T.Group />
      </TestCanvas>
    ));

    expect(ref!).toBeDefined();
  });

  it("should correctly unmount", async () => {
    let renderer = render(() => (
      <TestCanvas>
        <T.Group />
      </TestCanvas>
    ));

    expect(() => renderer.unmount()).not.toThrow();
  });

  // it("plays nice with react SSR", async () => {
  //   const useLayoutEffect = jest.spyOn(React, "useLayoutEffect");

  //   await act(async () => {
  //     render(
  //       <Canvas>
  //         <T.Group />
  //       </Canvas>,
  //     );
  //   });

  //   expect(useLayoutEffect).not.toHaveBeenCalled();
  // });
});
