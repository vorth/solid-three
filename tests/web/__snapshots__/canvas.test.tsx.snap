// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`web Canvas > should correctly mount 1`] = `
<div>
  <div
    style="width: 100%; height: 100%;"
  >
    <canvas
      data-engine="three.js r149"
      height="800"
      width="1280"
    />
  </div>
</div>
`;
